### ISSUE TITLE
Inferred array type allows unsound covariance?

### ISSUE THREAD

**Issue by jtbandes** [facebook/flow#8607]
The following code should raise an error, but doesn't:

[Try Flow](https://flow.org/try/#0MYewdgzgLgBAhjAvDA2gRgLoG4BQOBmArmMFAJbgz4ggAUcATgwFwy1iEC2ARgKYMAfaAzJgA5gEoUGCTADeOGEvhMAdAAdCEABa0ARAA89E3AF881OnBMXipCmBjdG9Jqw49+02QvM5nDPQmQA)
```js
const a = [1];

function foo(arr: (number|string)[]) {
    arr.push("x");
}
foo(a);

function bar(arr: number[]) { }
bar(a);
```

This problem seems to be specific to arrays. If I replace all `T[]` with `{x: T}`, then I get the expected behavior.

Flow version: v0.146.0

## Expected behavior

Either an error on `foo(a)`, or an error on `bar(a)`. If `a` is inferred to be `number[]`, then `foo(a)` should be an error. If it's inferred to be `(number|string)[]`, then `bar(a)` should be an error. 

## Actual behavior

When the type of the array is left off (inferred), it seems to produce unsound results. When the type is explicitly specified I get the expected results.

**Comment by mischnic** [comment]
It works as expected though when declaring the variable type explicitly.

```js
const a = [1];
const b: number[] = [1];

function foo(arr: (number|string)[]) {
    arr.push("x");
}

foo(a); // no error
foo(b); // error
```

**Comment by jtbandes** [comment]
Yes, I guess it’s a bug with the inference behavior.

**Comment by noppa** [comment]
Seems like the same thing as in #8108 and #8593

**Comment by jtbandes** [comment]
I agree—thanks for linking these. I think my repro case is a bit simpler since only primitive types are used 🙂

**Comment by SamChou19815** [comment]
Fixed in 0.278.0
